Retrieve diagnostics (for review purpose)

DaanHoogland opened the pull request.

charles-phiri committed:
Initial commit

charles-phiri committed:
some code to enhance the configure method in the Service Layer

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -44,21 +42,19 @@
-    HashMap<DiagnosticsKey.DiagnosticsType, Set<DiagnosticsKey<?>>> _diagnosticsTypeLevelsMap = new HashMap<DiagnosticsKey.DiagnosticsType, Set<DiagnosticsKey<?>>>();
+
+
+    \
```
what happened here? seems syntactically fine but kind of strange formatting

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -81,7 +77,7 @@ protected void populateConfiguration(Date date, Configurable configurable) {
         s_logger.debug("Retrieving keys from " + configurable.getClass().getSimpleName());
 
-        for (DiagnosticsKey<?> key : configurable..etConfigKeys()) {
+        for (DiagnosticsKey<?> key : configurable. . ..get.get.getConfigKeys()) {
```
some artifact that wasn't intended?

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsKey.java:
```diff
@@ -0,0 +1,139 @@
+
+
+
+/*    public ConfigKey.Scope scope() {
```
code in comment need not be, as we use a versioning system

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -14,16 +14,14 @@
 import org.apache.cloudstack.framework.config.ConfigDepot;
 import org.apache.cloudstack.framework.config.ConfigDepotAdmin;
-import org.apache.cloudstack.framework.config.impl.ConfigDepotImpl;
+import org.apache.cloudstack.framework.config.ScopedConfigStorage;
```
I cannot find this class. is it checked in?

DaanHoogland reviewed (commented):
nice concepts you use. I do not see the overall yet, but it looks promissing

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -16,12 +16,7 @@
-import org.apache.cloudstack.framework.config.ConfigDepot;
-import org.apache.cloudstack.framework.config.ConfigDepotAdmin;
-import org.apache.cloudstack.framework.config.ScopedConfigStorage;
+import org.apache.cloudstack.framework.config.*;
```
new one?

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -134,49 +124,37 @@ public RetrieveDiagnosticsDao global() {
     @Override
     public <T> void set(ConfigKey<T> key, T value) {
-        _configDao.update(key.key(), value.toString());
+        //_diagnosticsDao.update(key.key(), value.toString());
```
code in comment

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -134,49 +124,37 @@ public RetrieveDiagnosticsDao global() {
     @Override
     public <T> void createOrUpdateConfigObject(String componentName, ConfigKey<T> key, String value) {
-        createOrupdateConfigObject(new Date(), componentName, key, value);
+        //createOrupdateConfigObject(new Date(), componentName, key, value);
```
code in comment

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsKey.java:
```diff
@@ -53,23 +76,29 @@ static public void init(DiagnosticsConfigDepotImpl depot) {
-    public DiagnosticsKey(String role, DiagnosticsEntryType diagnosticstype, String description) {
-        _role = role;
-        _diagnosticsType = diagnosticstype;
+   /* public DiagnosticsKey(Class<T> type, String name, String description) {
```
code in comment

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsKey.java:
```diff
@@ -118,4 +147,8 @@ protected String valueOf(String value) {
         }
     }
 
+/*    public static DiagnosticsKey<?> getDiagnosticsClassKeys() {
```
code in comment

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -268,67 +283,88 @@ public boolean loadDiagnosticsConfiguration(final RetrieveDiagnosticsVO retrieve
-            //get list of default files from the database table for this diagnostics type
+                } else {
+                    throw new InvalidParameterValueException("Invalid parameters");
+                   // System.exit(-1);
```
code in comment

DaanHoogland reviewed (commented):
I like how you use the concept of generics. I also see what you are trying to achieve with it. I hope you are not making it to complicated for yourself. We are going to need a lot of unit and integration tests for this to have a clean end result.

charles-phiri committed:
refactored Service interface

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/RetrieveDiagnosticsVO.java:
```diff
@@ -52,18 +52,15 @@ public RetrieveDiagnosticsVO(String role, String className, String value) {
-
-    public RetrieveDiagnosticsVO(String roleId, String role, String className, String value) {
-        this.roleId = roleId;
+/*    public RetrieveDiagnosticsVO(String roleId, String role, String className, String value) {
```
no code in comment please

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/RetrieveDiagnosticsVO.java:
```diff
@@ -73,35 +70,26 @@ public void setRole(String role) {
     }
 
-    public void setClassName(String className) {
+    public void setDiagnosticsType(String className) {
```
:+1:

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsService.java:
```diff
@@ -35,28 +33,12 @@
 
 public interface RetrieveDiagnosticsService extends Manager, PluggableService {
 
-    ConfigKey<Long> RetrieveDiagnosticsTimeOut = new ConfigKey<Long>("Advanced", Long.class, "retrieveDiagnostics.retrieval.timeout", "3600",
```
why move to the Impl? isn't it fine here?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -161,15 +161,31 @@ public void set_configParams(Map<String, Object> _configParams) {
+            "The interval between garbage collection executions in seconds", true, ConfigKey.Scope.Global);
+
+
+/*    DiagnosticsKey<String> IPTablesRemove = new DiagnosticsKey<String>(String.class, "IPtables.remove", "The IPtables rules to be removed", null, null);
```
please remove code instead of commenting it out. we have a versioning system to retrieve old code.

DaanHoogland reviewed (commented):
the branch still doesn't compile for me.

charles-phiri committed:
refactored Service interface

charles-phiri committed:
Removed unused imports

charles-phiri committed:
refactored Service interface

charles-phiri committed:
refactored Service interface

charles-phiri committed:
This is a compiled version of the work I have been doing so far. I am busy testing the functionality and will make another commit for the working version of the work flow.

charles-phiri committed:
Changed the schema script which was having errors when deploying the DB.

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsService.java:
```diff
@@ -0,0 +1,44 @@
+
+    RetrieveDiagnosticsResponse getDiagnosticsFiles(final RetrieveDiagnosticsCmd cmd) throws InvalidParameterValueException, ConfigurationException;
+
+    ConfigKey<?>[] getConfigKeys();
```
this method shouldn't be declared here. It is part of interface Configurable.

charles-phiri replied:
Removed ConfigKey<?> getConfigKeys() from ServiceImpl

DaanHoogland replied:
It should remain in the ServiceImpl, @charles-phiri . But it shouldn't be in this interface. It is already defined in Configerable.

DaanHoogland commented on api/src/main/java/com/cloud/event/EventTypes.java:
```diff
@@ -266,6 +267,7 @@
     public static final String EVENT_SSVM_STOP = "SSVM.STOP";
     public static final String EVENT_SSVM_REBOOT = "SSVM.REBOOT";
     public static final String EVENT_SSVM_HA = "SSVM.HA";
+    public static final String EVENT_SSVM_DIAGNOSTICS = "SSVM.DIAGNOSTICS";
```
EVENT_SSVM_DIAGNOSTICS is not used outside this file. Do we really need it?

charles-phiri replied:
Removed EVENT_SSVM_DIAGNOSTICS

DaanHoogland commented on engine/schema/src/main/resources/META-INF/db/schema-41110to41200.sql:
```diff
@@ -32,4 +32,27 @@ ALTER TABLE `vlan` CHANGE `description` `ip4_range` varchar(255);
\ No newline at end of file
+INSERT INTO `cloud`.`role_permissions` (`uuid`, `role_id`, `rule`, `permission`, `sort_order`) values (UUID(), 4, 'moveNetworkAclItem', 'ALLOW', 260) ON DUPLICATE KEY UPDATE rule=rule;
+
+CREATE TABLE `cloud`.`diagnosticsdata` (
```
isn't this more of a 'diagnostics_defaults' or 'diagnostics_configuration'? I don't think the name diagnosticsdata is right for this table.

charles-phiri replied:
I think it is a matter of preference. The name was suggested by Dag....don't know what you think?

DaanHoogland replied:
diagnostics data to me suggests that the result of any action or gathering is in this table, as opposed to it containing a means of steering how gathering is done. Hence my two suggestions. I am not religious obout those two, but I think the name diagnosticsdata is misleading.

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -0,0 +1,167 @@
+
+    private void createOrupdateDiagnosticsObject(String componentName,  DiagnosticsKey<?> diagnosticsType) {
+        RetrieveDiagnosticsVO vo = _diagnosticsDao.findById(diagnosticsType.key());
+        //DiagnosticsKey diagnosticsKey = new DiagnosticsKey(diagnosticsType.getClass(), diagnosticsType.key(), "new diagnostics")
```
code in comment

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -0,0 +1,167 @@
+
+    private final static Logger s_logger = Logger.getLogger(DiagnosticsConfigDepotImpl.class);
+    @Inject
+    RetrieveDiagnosticsDao _diagnosticsDao;
```
we don't need the antiquated convention of prefix '_' on member names.

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -0,0 +1,167 @@
+
+    @Override
+    public <T> void set(ConfigKey<T> key, T value) {
+        //_diagnosticsDao.update(key.key(), value.toString());
```
code in comment

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -0,0 +1,167 @@
+
+    @Override
+    public <T> void createOrUpdateConfigObject(String componentName, ConfigKey<T> key, String value) {
+        //createOrupdateConfigObject(new Date(), componentName, key, value);
```
code in comment

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsKey.java:
```diff
@@ -0,0 +1,164 @@
+
+public class DiagnosticsKey<T> {
+    public static enum DiagnosticsEntryType {
+        IPTABLES, LOGFILES, PROPERTYFILES, DHCPFILES, USERDATA, LB, DNS, VPN, IPTABLESretrieve, IPTABLESremove
```
is there a reason why most are all caps and the last two are mixed case?

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsKey.java:
```diff
@@ -0,0 +1,164 @@
+        s_depot = depot;
+    }
+
+   /* public DiagnosticsKey(Class<T> type, String name, String description) {
```
code in comment

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsKey.java:
```diff
@@ -0,0 +1,164 @@
+        }
+    }
+
+/*    public static DiagnosticsKey<?> getDiagnosticsClassKeys() {
```
code in comment

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/RetrieveDiagnosticsVO.java:
```diff
@@ -0,0 +1,111 @@
+        this(key.key(), key.getClassName(), key.getDetail());
+    }
+
+/*    public RetrieveDiagnosticsVO(String roleId, String role, String className, String value) {
```
code in comment

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,409 @@
+
+                } else {
+                    throw new InvalidParameterValueException("Invalid parameters");
+                   // System.exit(-1);
```
code in comment

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,409 @@
+
+    public DiagnosticsKey<?>[] getDiagnosticsConfigKeys()
+    {
+        return null; //new DiagnosticsKey<?>[] { IPTablesRemove, IPTablesRetrieve, LOGFILES, PROPERTYFILES, DNSFILES, DHCPFILES, USERDATA, LB, VPN   };
```
code in comment

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,409 @@
+        return cmdList;
+    }
+
+/*    public SecondaryStorageVmVO startSecondaryStorageVm(long secStorageVmId) {
```
code in comment

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,409 @@
+                if (fileDetails != null) {
+                    filesToRetrieve = fileDetails.split(",");
+                } else {
+                    filesToRetrieve = null;//retrieve default files from db
```
is this comment a TODO?

DaanHoogland commented on tools/apidoc/gen_toc.py:
```diff
@@ -190,7 +190,8 @@
     'cloudian': 'Cloudian',
-    'Sioc' : 'Sioc'
+    'Sioc' : 'Sioc',
+    'retrieveDiagnostics': 'System VM'
```
is this the right tag/token? it seems to me that it deserves its own class. It will also be applicable to hosts in the future (maybe)

charles-phiri replied:
From the documentation, it looks to me like it is the right file.

DaanHoogland replied:
yes, it is the right file. I wonder about the value of the tag, not the place it is defined

DaanHoogland reviewed (commented):
I had a new high-level pass on the whole of it. I have two main questions:
1. naming in general needs attention. Can we improve on that a bit.
2. I want to see why you created the generic. What is your purpose for it?

charles-phiri committed:
Minor fixes on SQL schema file

charles-phiri committed:
Changed the schema script which was having errors when deploying the DB.

DaanHoogland commented on core/src/main/java/com/cloud/agent/api/RetrieveDiagnosticsAnswer.java:
```diff
@@ -0,0 +1,59 @@
+/*
+ * // Licensed to the Apache Software Foundation (ASF) under one
```
please remove the nested comment

charles-phiri replied:
Done

DaanHoogland commented on core/src/main/java/com/cloud/agent/api/RetrieveDiagnosticsAnswer.java:
```diff
@@ -0,0 +1,59 @@
+ * // under the License.
+ */
+
+// Licensed to the Apache Software Foundation (ASF) under one
```
three licenses is a bit overdone

DaanHoogland commented on core/src/main/java/com/cloud/agent/api/RetrieveDiagnosticsCommand.java:
```diff
@@ -0,0 +1,90 @@
+/*
+ * // Licensed to the Apache Software Foundation (ASF) under one
```
same issue as in answer

charles-phiri replied:
Done

charles-phiri committed:
Refactored some code in the API (Cmd) class.

charles-phiri committed:
Added some more functionality in the Service Layer class.

charles-phiri committed:
Removed the code that was using generics in the class DiagnosticsKey.

charles-phiri committed:
Corrected code in my DiagnosticsConfigDepotImpl class that was using the generics code.

charles-phiri committed:
Corrected code in my RetrieveDiagnosticsVO class that was using the generics code.

charles-phiri committed:
Yet still, removed some code using generics code.

charles-phiri committed:
Added more code in the Service Layer class and refactored code in the Dao and DiagnosticsKey classes.

charles-phiri committed:
Refactored code in Service Impl class.

charles-phiri committed:
Added code in spring xml file.

charles-phiri committed:
Created my own interface class for DiagnosticsConfigDepotImpl as I could not @Inject because of the intreference from ConfigDepot which implements the same interfaces.

charles-phiri committed:
Changed most of the code in the DaO and VO classes including its usage in the Service Layer.

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/api/command/admin/diagnostics/RetrieveDiagnosticsCmd.java:
```diff
@@ -63,95 +58,73 @@
 
-    public RetrieveDiagnosticsService getRetrieveDiagnosticsService() {
-        return retrieveDiagnosticsService;
+    public static Logger getS_logger() {
```
why an accessor for the logger? it is only used locally isn't it?

DaanHoogland replied:
also it should be called LOGGER or LOG as it is static final. and not have the obsoleted 's_' prefix

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/api/command/admin/diagnostics/RetrieveDiagnosticsCmd.java:
```diff
@@ -63,95 +58,73 @@
 
-    public void setRetrieveDiagnosticsService(RetrieveDiagnosticsService retrieveDiagnosticsService) {
-        this.retrieveDiagnosticsService = retrieveDiagnosticsService;
+    public static String getAPINAME() {
```
this is not to be accessed outside the class is it?

DaanHoogland commented on engine/schema/src/main/resources/META-INF/db/schema-41110to41200.sql:
```diff
@@ -40,15 +40,15 @@ CREATE TABLE `cloud`.`diagnosticsdata` (
-INSERT INTO `cloud`.`diagnosticsdata` (`role`, `class`, `value`) values ('VR', 'VPN', '<vpn configuration file>');
-INSERT INTO `cloud`.`diagnosticsdata` (`role`, `class`, `value`) values ('VR', 'LOGFILES', 'cloud.log,agent.log');
-INSERT INTO `cloud`.`diagnosticsdata` (`role`, `class`, `value`) values ('CPVM', 'PROPERTYFILES', '<CPVM property file>');
+INSERT INTO `cloud`.`diagnosticsdata` (`role`, `class`, `value`) values ('SecondaryStorageVm', 'LOGFILES', 'cloud.log,agent.log,[IPTABLES]');
```
:+1:

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/DiagnosticsConfigDepot.java:
```diff
@@ -0,0 +1,38 @@
+/*
```
can you remove the nesting in the comment, please?

charles-phiri replied:
Done.

DaanHoogland replied:
no, it is still here

charles-phiri replied:
I have changed it. Will make a push soon after completing the code I am working on.

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/RetrieveDiagnosticsDaoImpl.java:
```diff
@@ -0,0 +1,62 @@
+/*
+ * // Licensed to the Apache Software Foundation (ASF) under one
```
please remove the nesting in the comment.

charles-phiri replied:
Done.

DaanHoogland replied:
but you didn't commit it?

charles-phiri committed:
Refactored code in the Dao and VO, and also added one more method to retrieve all the data and store it in a HashMap.

charles-phiri committed:
Cleaned up the code and refactored code so that it has some of the changes discussed in the mentor review.

charles-phiri committed:
Refactored code in the Service Layer impl.

DaanHoogland reviewed (commented):
so far, now.
pleae push further changes asap.

charles-phiri committed:
Made some code logic changes to the Service Layer when populating the dynamic classes and made a few more clean ups of the code.

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/api/response/RetrieveDiagnosticsResponse.java:
```diff
@@ -37,6 +37,11 @@
     private Date lastSent;
 
+    @SerializedName(ApiConstants.TIMEOUT)
+    @Param(description = "the timeout (in seconds) for requests to the retrieve diagnostics API")
```
what is the purpose of this in the response? i.e. what information does the user get from this response field?

charles-phiri committed:
Addressed comments on the code.

charles-phiri committed:
Refactored code in the Service Layer for retrieving recodes from the database table.

DaanHoogland commented on api/src/main/java/com/cloud/event/EventTypes.java:
```diff
@@ -740,6 +743,7 @@
         entityEventDetails.put(EVENT_SSVM_STOP, VirtualMachine.class);
         entityEventDetails.put(EVENT_SSVM_REBOOT, VirtualMachine.class);
         entityEventDetails.put(EVENT_SSVM_HA, VirtualMachine.class);
+        entityEventDetails.put(EVENT_SSVM_DIAGNOSTICS, VirtualMachine.class);
```
when is this fired?

charles-phiri replied:
Removed it but I think we might need Events at a later stage.

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/api/command/admin/diagnostics/RetrieveDiagnosticsCmd.java:
```diff
@@ -0,0 +1,159 @@
+        RetrieveDiagnosticsResponse retrieveDiagnosticsResponse = new RetrieveDiagnosticsResponse();
+        try {
+            retrieveDiagnosticsService.getDiagnosticsFiles(this);
+            retrieveDiagnosticsResponse.setObjectName("retrievediagnostics");
```
is this the ideal object name? did you have other names considdered?

charles-phiri replied:
What can you suggest?

DaanHoogland replied:
"diagnostics" maybe? or retrievedlogs, diagnosticsinformation. diagnosticdata.
I am really not sure it is a sincere question. It is just that retrievediagnostics (which is a spelling error btw, only one d) does not seem right. 
from wiktionary: diagnostics

    plural of diagnostic
    The process of determining the state of or capability of a component to perform its function(s).
retrieving a process is not intuitively right.

charles-phiri replied:
I think diagnosticsinformation makes sense. Do you want me to change the whole class or just retrieveDiagnosticsResponse.setObjectName("retrievediagnostics");

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/api/command/admin/diagnostics/RetrieveDiagnosticsCmd.java:
```diff
@@ -0,0 +1,159 @@
+
+    @Override
+    public String getEventDescription() {
+        return "Retrieved diagnostics files from System VM =" + id;
```
"System VM ="  may be too specific if we add hosts or hardware loadbalancers. We might go for "host: ". What do you think?

charles-phiri replied:
Perfectly makes sense

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/api/command/admin/systemvm/DestroySystemVmCmd.java:
```diff
@@ -16,8 +16,9 @@
 // under the License.
 package org.apache.cloudstack.api.command.admin.systemvm;
 
-import org.apache.log4j.Logger;
```
there is only imports chaged in this file. consider reverting it.

charles-phiri replied:
Okay

charles-phiri replied:
Reverted it.

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsService.java:
```diff
@@ -0,0 +1,49 @@
+
+    ConfigKey<?>[] getConfigKeys();
+
+    boolean configure(final String name, final Map<String, Object> params) throws ConfigurationException;
```
should this be in the interface?

DaanHoogland replied:
Charles, why is this in the interface?

DaanHoogland commented on core/src/main/java/com/cloud/agent/api/RetrieveDiagnosticsCommand.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @Override
+    public boolean isQuery() {
```
please look at the formatting here

charles-phiri replied:
Okay

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/RetrieveDiagnosticsDaoImpl.java:
```diff
@@ -0,0 +1,69 @@
+@Component
+public class RetrieveDiagnosticsDaoImpl extends GenericDaoBase<RetrieveDiagnosticsVO, String> implements RetrieveDiagnosticsDao
+{
+    private final SearchBuilder<RetrieveDiagnosticsVO> DiagnosticsSearchByType;
```
can you check if we need this, please. it is not used in the class or in its interface. I think you blindly copied my code from annotations and I should have removed it there myself. I am not sure, though.

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -0,0 +1,103 @@
+import java.util.HashMap;
+import java.util.List;
+
+public class DiagnosticsConfigDepotImpl implements DiagnosticsConfigDepot {
```
The name depot is not very fortunate. Please help think of a better one. I am thinking of just DiagnosticsConfiguration or DiagnosticsConfigurator.

charles-phiri replied:
You are right, thanks

charles-phiri replied:
Done

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/DiagnosticsConfigDepotImpl.java:
```diff
@@ -0,0 +1,103 @@
+
+public class DiagnosticsConfigDepotImpl implements DiagnosticsConfigDepot {
+
+    private final static Logger s_logger = Logger.getLogger(DiagnosticsConfigDepotImpl.class);
```
the naming convention of s_ is obsolete, please use the name for a static final in all uppercase; i.e. LOG or LOGGER

charles-phiri replied:
Done. Will have to look at other classes and change as required.

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/DiagnosticsConfigDepot.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.HashMap;
+import java.util.List;
+
+public interface DiagnosticsConfigDepot {
```
naming is a bit off. please see the -Impl for discussion

charles-phiri replied:
We have to discuss this so that I can learn the naming conversion. Meanwhile, I will leave it as it is.

DaanHoogland replied:
I think you mean convention, right? I am not sure there is a convention for what you are trying to do here yet. That would mean we are free to choose. Of course with that freedom comes great responsibility ;) I must be intuitively right.

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/api/command/admin/diagnostics/RetrieveDiagnosticsCmd.java:
```diff
@@ -55,10 +54,9 @@
     @Parameter(name = ApiConstants.ID,
             type = CommandType.UUID,
-            entityType = RetrieveDiagnosticsResponse.class,
+            entityType = DomainRouterResponse.class,
```
So this call is now only valid for DomainRouters? using the DomainRouterResponse.class...? I think something went wrong here

charles-phiri committed:
Added unit test file to test the functionality in the Service Layer implementation.

charles-phiri committed:
Added unit test file to test the functionality in the Service Layer implementation.

charles-phiri committed:
Removed the duplicated licence information.

charles-phiri committed:
1. Cleaned up code as per Daan's comments.
2. Reverted the DestroySystemVmCmd.java file.
3. Added some code to select Secondary Storage.

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/RetrieveDiagnosticsDaoImpl.java:
```diff
@@ -31,14 +31,14 @@
 public class RetrieveDiagnosticsDaoImpl extends GenericDaoBase<RetrieveDiagnosticsVO, String> implements RetrieveDiagnosticsDao
 {
-    private final SearchBuilder<RetrieveDiagnosticsVO> DiagnosticsSearchByType;
+//    private final SearchBuilder<RetrieveDiagnosticsVO> DiagnosticsSearchByType;
```
did it work, without? (just remove, git will find it back if you need it;)

charles-phiri replied:
Yes, it did work thanks

charles-phiri committed:
Mgt-server could be running in a vm that does not have diskOffering. Just added code to check for that before it can save the tar file from a System VM.

charles-phiri committed:
Added code to check capacity of the disk space and compare it to the threshold.

charles-phiri committed:
Added code to check capacity of the disk space and compare it to the threshold.

charles-phiri committed:
Added code to check capacity of the disk space and compare it to the threshold.

DaanHoogland commented on core/src/main/java/com/cloud/agent/api/RetrieveDiagnosticsAnswer.java:
```diff
@@ -0,0 +1,24 @@
+
+/*
+ * // Licensed to the Apache Software Foundation (ASF) under one
```
this is still nested comment

charles-phiri replied:
Done

charles-phiri committed:
Added code to check capacity of the disk space and compare it to the threshold.Also added a skeleton script.

charles-phiri committed:
Added code to check capacity of the disk space and compare it to the threshold.Also added a skeleton script.

charles-phiri committed:
Added code to parse for a default value which is in square brackets, and then creating a name of a script to be run.

charles-phiri committed:
Inserted getters/setters for the script filename variables

charles-phiri committed:
Added skeleton scripts (still adding code) to retrieve files and run ip route commands, iptables command and ifconfig commands. Also added two separate Command classes for execution of the scripts and for file retrieval.

charles-phiri committed:
Added code for the two command classes (ExecuteScriptCommand and RetrieveFileCommand) and logic to parse for the scripts and files.

charles-phiri committed:
Added code for the two command classes (ExecuteScriptCommand and RetrieveFileCommand) and logic to parse for the scripts and files.

charles-phiri committed:
Added code for the two command classes (ExecuteScriptCommand and RetrieveFileCommand) and logic to parse for the scripts and files.

charles-phiri committed:
Add method to check disk space

charles-phiri committed:
Minor changes in unit test

charles-phiri committed:
Still working on unit test

charles-phiri committed:
Added more unit tests.

charles-phiri committed:
Skeleton implementation of the Garbage Collector.

borisstoyanov commented on systemvm/debian/opt/cloud/bin/retrieve_diagnostics.py:
```diff
@@ -0,0 +1,130 @@
+
+
+
+
```
is there any reason we need that many empty lines?

borisstoyanov reviewed (commented):
Can we also add a marvin test?

charles-phiri committed:
Refactored code while testing, mostly in the Service Layer. Added a copyFileFromSystemVm method in the Service Layer to copy files from System VM using the SshHelper class

charles-phiri committed:
Got rid of the cyclic dependencies that I was encountering when compiling after the implementation of the RetrieveDiagnostics API Garbage Collector - DiagnosticsGarbageCollector

charles-phiri committed:
Merge branch 'master' of https://github.com/apache/cloudstack into retrieve-diagnostics

charles-phiri committed:
Refactored code in the Service Layer and the python scripts for retrieving diagnostics files from the System VM.

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/diagnostics/DeleteZipCommand.java:
```diff
@@ -0,0 +1,28 @@
+import com.cloud.agent.api.Command;
+import com.cloud.agent.api.to.DataStoreTO;
+
+public class DeleteZipCommand extends Command {
```
I know it is not custm in cloudstack but can you add some documentation?
i.e:
```
/**
 * command to clean up after retrieving diagnostics files
 */
```

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/diagnostics/DeleteZipCommand.java:
```diff
@@ -0,0 +1,28 @@
+import com.cloud.agent.api.to.DataStoreTO;
+
+public class DeleteZipCommand extends Command {
+    private String zipFile;
```
maybe name it zipFileName?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -92,29 +97,41 @@
 
     private static final Logger LOGGER = Logger.getLogger(RetrieveDiagnosticsServiceImpl.class);
 
+    protected StorageLayer _storage;
```
_ on fields are obsolete, as this file is already mixed, please remove.

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -126,6 +143,12 @@
 
+    private Integer nfsVersion;
+
+    protected String _parent = "/mnt/SecStorage";
```
_

DaanHoogland replied:
please remove the underscore from the name

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -600,8 +622,27 @@ public DiagnosticsGarbageCollector() {
                 LOGGER.trace("Diagnostics Garbage Collection Thread is running.");
-
-                cleanupDiagnostics();
+                File zipFile = new File(path + zipFileName);
```
isn't zipFileName always null at this moment?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -600,8 +622,27 @@ public DiagnosticsGarbageCollector() {
+                try {
+                    attributes = Files.readAttributes(zipFilePath, BasicFileAttributes.class);
+                } catch (IOException exception) {
+                    LOGGER.error("Could not find the zip file.", exception);
```
why is it an error when there is nothing to clean up?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -610,7 +651,212 @@ public void run() {
 
     @Override
-    public void cleanupDiagnostics() {
+    public Answer cleanupDiagnostics(DeleteZipCommand cmd) {
```
Is this supposed to run in the server or in the agent?
if in the agent it shouldn't be implemented here.
if on the management server there is no need for the Command/Answer pattern
Can you explain your thinking here?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -610,7 +651,212 @@ public void run() {
+
+        attemptMount(localRootPath, remoteDevice, uri, nfsVersion);
+
+        checkForVolumesDir(localRootPath);
```
check for volumes inside of mount; that feels like a side effect.

DaanHoogland replied:
the return of the method is ignored. why is the return implemented?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -610,7 +651,212 @@ public void run() {
+        checkForVolumesDir(localRootPath);
+    }
+
+    protected boolean checkForVolumesDir(String mountPoint) {
```
checkForVolumesDir is not checking but actually creating. beter name?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -610,7 +651,212 @@ public void run() {
+                LOGGER.debug(dirName + " already exists on secondary storage, and is mounted at " + mountPoint);
+                dirExists = true;
+            } else {
+                if (dir.delete() && _storage.mkdir(dirLocation)) {
```
further down you don't assume this succeeds, why do you here?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -610,7 +651,212 @@ public void run() {
+            _storage.mkdir(file.getPath());
+
+            // Need to check after mkdir to allow O/S to complete operation
+            if (!file.exists()) {
```
above you assume mkdir succeeds but not here. Why?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -610,7 +651,212 @@ public void run() {
+        }
+
+        if (dirExists) {
+            LOGGER.info(dirName + " directory created/exists on Secondary Storage.");
```
you have obviously given log levels thoughts. let's discuss them. As it seems these two might not be of the same level.

DaanHoogland reviewed (commented):
some comments
code looks fine but i fear for architectural fit at some places.

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,1048 @@
+    }
+
+    @Override
+    public Answer cleanupDiagnostics(DeleteZipCommand cmd) {
```
Is this supposed to run in the server or in the agent?
if in the agent it shouldn't be implemented here.
if on the management server there is no need for the Command/Answer pattern
Can you explain your thinking here?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,1048 @@
+        public void run() {
+            try {
+                LOGGER.trace("Diagnostics Garbage Collection Thread is running.");
+                File zipFile = new File(path + zipFileName);
```
isn't zipFileName always null at this moment?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,1048 @@
+
+    private static final Logger LOGGER = Logger.getLogger(RetrieveDiagnosticsServiceImpl.class);
+
+    protected StorageLayer _storage;
```
please remove underscores on fields. it is an old obsoleted convention

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/DiagnosticsConfigurator.java:
```diff
@@ -0,0 +1,39 @@
+/*
+ * // Licensed to the Apache Software Foundation (ASF) under one
```
please remove the nesting in these comments

charles-phiri committed:
Marvin tests for the RetrieveDiagnostics API

borisstoyanov commented on test/integration/smoke/test_retrieve_diagnostics.py:
```diff
@@ -0,0 +1,235 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
As of structure I think this looks good, let me get back on this when I pick this up for testing and possibly point any missing scenario.

charles-phiri committed:
Full complete code after refactoring. Addition of copyFileFromSystemVmToServer and downloadCompressedFileToSSVM methods in the Service Layer.

DaanHoogland commented on api/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsService.java:
```diff
@@ -38,13 +37,13 @@
     List<DiagnosticsKey> get(String key);
 
-    Answer cleanupDiagnostics(DeleteZipCommand cmd);
+//    boolean cleanupDiagnostics(DeleteCommand cmd);
```
code in comment, not needed (git)

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -410,159 +465,204 @@ protected String getDefaultFilesForVm(String diagnosticsType, String systemVmTyp
+            throw new CloudRuntimeException("Unlimited disk space on primary storage.");
+        }
+        String publicIp = null;
+        //final Map<String, String> accessDetails = networkManager.getSystemVMAccessDetails(vmInstance);
```
please delete commented code

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -410,159 +465,204 @@ protected String getDefaultFilesForVm(String diagnosticsType, String systemVmTyp
+        for (final NicVO nic : nics) {
+            final NetworkVO network = _networksDao.findById(nic.getNetworkId());
+            final NicVO nicVO = _nicDao.findByNtwkIdAndInstanceId(network.getId(), vm.getId());
+            publicIp = nicVO.getIPv4Address();
```
maybe break when found, no need to iterate over the rest if we are happy with an access point

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -410,159 +465,204 @@ protected String getDefaultFilesForVm(String diagnosticsType, String systemVmTyp
+            String copyCommand = new String("scp root@" + vmInstance.getPrivateIpAddress() + ":/temp/diagnostics_* " + "/diagnosticsdata");
+            LOGGER.info("Executing " + copyCommand);
+            sess.execCommand(copyCommand);
+            Thread.sleep(120000);
```
I don't trust this sleep. How do we know if it is enough or maybe excessive?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -410,159 +465,204 @@ protected String getDefaultFilesForVm(String diagnosticsType, String systemVmTyp
+            sess.close();
+            //close the connection
+            conn.close();
+        } catch (Exception ex) {
```
can we be more specific about what we catch (instead of pokemon catching)?

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -410,159 +465,204 @@ protected String getDefaultFilesForVm(String diagnosticsType, String systemVmTyp
 
-    protected Answer copyFileFromSystemVm(final NetworkElementCommand cmd, String script) throws Exception {
-        final File keyFile = getSystemVMKeyFile();
+    private CreateEntityDownloadURLAnswer downloadCompressedFileToSSVM(CreateEntityDownloadURLCommand cmd, final String diagnosticsZipFileName) {
```
can you break this up in smaller methods?
i.e. checkApache(), mkdirCommand(), linkCommand() etc...

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -144,12 +154,24 @@
+    private String secUrl;
+
+    @Inject
+    NetworkModel _networkModel;
```
please remove the underscore from the name

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -144,12 +154,24 @@
+    NetworkModel _networkModel;
+
+    @Inject
+    NetworkDao _networksDao = null;
```
please remove the underscore from the name

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -144,12 +154,24 @@
+    NetworkDao _networksDao = null;
+
+    @Inject
+    NicDao _nicDao = null;
```
please remove the underscore from the name

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -144,12 +154,24 @@
+    NicDao _nicDao = null;
 
     @Inject
     private HostDao _hostDao;
```
please remove the underscore from the names, here and below

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -220,6 +239,11 @@
     private VirtualMachineManager vmManager;
 
+    @Inject
+    protected DataCenterDao _dcDao = null;
```
please remove the underscore from the name

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -220,6 +239,11 @@
+    protected DataCenterDao _dcDao = null;
+
+    @Inject
+    protected PrimaryDataStoreDao _storagePoolDao = null;
```
please remove the underscore from the name

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -341,6 +364,8 @@ protected void loadDiagnosticsDataConfiguration() {
         }
         final Long vmId = cmd.getId();
         vmInstance = _vmDao.findByIdTypes(vmId, VirtualMachine.Type.ConsoleProxy, VirtualMachine.Type.DomainRouter, VirtualMachine.Type.SecondaryStorageVm);
+        //vmInstance.getDetails()
```
please remove code in comment

DaanHoogland commented on core/src/main/java/com/cloud/agent/api/RetrieveFilesCommand.java:
```diff
@@ -0,0 +1,52 @@
+/*
+ * // Licensed to the Apache Software Foundation (ASF) under one
```
this is a nested comment, please use either /* */ or //

DaanHoogland commented on framework/config/src/main/java/org/apache/cloudstack/framework/config/impl/RetrieveDiagnostics.java:
```diff
@@ -0,0 +1,29 @@
+/*
+ * // Licensed to the Apache Software Foundation (ASF) under one
```
please remove nested comment

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,1127 @@
+    NicDao _nicDao = null;
+
+    @Inject
+    private HostDao _hostDao;
```
please remove the _ from this field and all those below

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,1127 @@
+                filesToRetrieve.add(files[i]);
+            }
+        }
+       // ManagementServerHostVO managementServerHostVO = managementServerHostDao.findById(systemVmId..getHostId());
```
remove commented code

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,1127 @@
+            }
+        }
+       // ManagementServerHostVO managementServerHostVO = managementServerHostDao.findById(systemVmId..getHostId());
+        //String ipHostAddress = managementServerHostVO.getServiceIP();
```
remove commented code

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,1127 @@
+        RetrieveFilesCommand retrieveFilesCommand = null;
+        Map<String, String> resultsMap;
+        final Long hostId = systemVmId.getHostId();
+        //Answer answer = null;
```
remove commented code

DaanHoogland commented on server/src/main/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImpl.java:
```diff
@@ -0,0 +1,1127 @@
+                if (store == null) {
+                    throw new CloudRuntimeException("cannot find an image store for zone " + zoneId);
+                }
+                //DataStoreTO destStoreTO = store.getTO();
```
remove commented code

DaanHoogland commented on server/src/test/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImplTest.java:
```diff
@@ -0,0 +1,228 @@
+/*
+ * // Licensed to the Apache Software Foundation (ASF) under one
```
remove nested comments

DaanHoogland commented on server/src/test/java/org/apache/cloudstack/diagnostics/RetrieveDiagnosticsServiceImplTest.java:
```diff
@@ -0,0 +1,228 @@
+
+        }
+
+        Map<String, String> resultsMap = new HashMap<>();//retrieveDiagnosticsService.getDiagnosticsFiles(retrieveDiagnosticsCmd);
```
remove the commented code at the end of this line

DaanHoogland commented on systemvm/debian/opt/cloud/bin/retrieve_diagnostics.py:
```diff
@@ -0,0 +1,126 @@
+            except OSError as e:
+                print("Failed to create directory temp")
+
+        #manifest_file = open("temp/" + "manifest.txt", "w+")
```
remove commented code

DaanHoogland commented on systemvm/debian/opt/cloud/bin/retrieve_diagnostics.py:
```diff
@@ -0,0 +1,126 @@
+                print filename + " not found"
+                #manifest_file.write(filename + " not found.")
+
+        #zip_archive.write(manifest_file)
```
remove commented code

DaanHoogland commented on systemvm/debian/opt/cloud/bin/retrieve_diagnostics.py:
```diff
@@ -0,0 +1,126 @@
+                #manifest_file.write(filename + " not found.")
+
+        #zip_archive.write(manifest_file)
+        #manifest_file.close()
```
remove commented code

DaanHoogland commented on systemvm/debian/opt/cloud/bin/iptables.py:
```diff
@@ -0,0 +1,67 @@
+    save_files = SaveIptablesToLogFile(arguments)
+    save_files.ensure_dir(file_path)
+    save_files.saveIpTableEntries(file_path)
+
```
too many extra lines

DaanHoogland commented:
@charles-phiri please rebase to resolve conflicts and make sure the build passes

charles-phiri committed:
Code cleanup and refactoring

charles-phiri committed:
Complete RerieveDiagnostics API implementation for testing.

charles-phiri committed:
Complete RerieveDiagnostics API .

charles-phiri committed:
Minor bug fixes.

charles-phiri committed:
Code for the RetrieveDiagnostics API.
